From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 1 May 2022 19:20:12 +0900
Subject: [PATCH] Purpur PR - Add toggle for enchant level clamping
 (PurpurMC/Purpur#966)

Original by 12emin34 <macanovic.emin@gmail.com>

Copyright (c) 2022 PurpurMC LLC

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index a999e4c82d7fe0ecf127f9132c29886d56c444e5..32777f56ee562c158740960dc1555c7dd0b23ea0 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -1151,7 +1151,7 @@ public final class ItemStack {
 
         ListTag nbttaglist = this.tag.getList("Enchantments", 10);
 
-        nbttaglist.add(EnchantmentHelper.storeEnchantment(EnchantmentHelper.getEnchantmentId(enchantment), (byte) level));
+        nbttaglist.add(EnchantmentHelper.storeEnchantment(EnchantmentHelper.getEnchantmentId(enchantment), (net.prismarineteam.prismarine.PrismarineConfig.clampEnchantLevels) ? (byte) level : ((net.prismarineteam.prismarine.PrismarineConfig.useInteger) ? (short) level : (int) level))); // Prismarine - Purpur PR
         processEnchantOrder(this.tag); // Paper
     }
 
diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index b3292c67658659cd3cf5117d70e20921db3a1e21..bbf098dfe1991b093aa3d06efc3a2a5cd6ac6134 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -46,7 +46,7 @@ public class EnchantmentHelper {
     }
 
     public static int getEnchantmentLevel(CompoundTag nbt) {
-        return Mth.clamp(nbt.getInt("lvl"), 0, 255);
+        return Mth.clamp(nbt.getInt("lvl"), 0, (net.prismarineteam.prismarine.PrismarineConfig.clampEnchantLevels) ? 255 : ((net.prismarineteam.prismarine.PrismarineConfig.useInteger) ? 32767 : 2147483647)); // Prismarine - Purpur PR
     }
 
     @Nullable
diff --git a/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java b/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java
index 4dad7fc69bd213fcd20c24b032e1fa70cc6d0366..bfbc46b98e3fd39671074da9abbb5599716e5ced 100644
--- a/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java
+++ b/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java
@@ -155,4 +155,11 @@ public class PrismarineConfig {
     private static void useAirplaneCollisionContext() {
         useNormalCollisionContext = getBoolean("settings.use-normal-collision-context", useNormalCollisionContext);
     }
+
+    public static boolean clampEnchantLevels = false;
+    public static boolean useInteger = false;
+    private static void purpurPRs(){
+        clampEnchantLevels = getBoolean("settings.pull-requests.purpur.clamp-enchantment-levels", clampEnchantLevels);
+        clampEnchantLevels = getBoolean("settings.pull-requests.purpur.clamp-enchantment-levels.use-integer", useInteger);
+    }
 }
\ No newline at end of file
